### ISSUE TITLE
MacOS: The mouse wheel needs several clicks to select prev/next weapon.

### ISSUE THREAD

**Issue by Velociraptor555** [ioquake/ioq3#379]
The problem has been around for some time, I always pull the latest version from git and compile it prior to start the game. I used the latest macOS (High Sierra), Xcode and command line tools.

It is really hard to play the game that way. I've compiled the latest git revision to a 64 bit Windows binary: No problem on changing weapons with the mouse wheel there.

It would be nice if there was a fix soon.

**Comment by ensiform** [comment]
What version of sdl is being used do you know off hand?

**Comment by Velociraptor555** [comment]
otool -L ioquake3  | grep -i sdl

yields

@executable_path/libSDL2-2.0.0.dylib (compatibility version 9.0.0, current version 9.0.0)

**Comment by ensiform** [comment]
I don't think that tells the version compiled.  It's always called 2.0.0 regardless on Mac.

**Comment by Velociraptor555** [comment]
port list installed | grep -i sdl

outputs

libsdl                         @1.2.15         devel/libsdl
libsdl2                        @2.0.8          devel/libsdl2

**Comment by ensiform** [comment]
So the only two possibilities I can think of at the moment is incorrect handling of the mouse wheel in the ioquake3 translator code for SDL or it could be an actual SDL bug.

**Comment by mcaldas** [comment]
I am having this issue too, on mac with latest build. Any idea how to fix it?

**Comment by zturtleman** [comment]
To help diagnose whether this is an issue in ioquake3 or SDL, edit code/sdl/sdl_input.c to add this Com_Printf message and then recompile ioquake3 and check if each scroll click prints one message to the console.
```
static void IN_ProcessEvents( void )
{
...
			case SDL_MOUSEWHEEL:
				// new debug message
				Com_Printf( "Mouse wheel event: x: %d, y: %d\n", e.wheel.x, e.wheel.y );

				if( e.wheel.y > 0 )
				{
					Com_QueueEvent( in_eventTime, SE_KEY, K_MWHEELUP, qtrue, 0, NULL );
					Com_QueueEvent( in_eventTime, SE_KEY, K_MWHEELUP, qfalse, 0, NULL );
				}
				else if( e.wheel.y < 0 )
				{
					Com_QueueEvent( in_eventTime, SE_KEY, K_MWHEELDOWN, qtrue, 0, NULL );
					Com_QueueEvent( in_eventTime, SE_KEY, K_MWHEELDOWN, qfalse, 0, NULL );
				}
				break;
```

**Comment by Velociraptor555** [comment]
Result with instrumented IN_ProcessEvents(void):

1. After start up, every wheel "klick" prints a line to console, but doesn't change weapon.
2. After a very fast wheel move, console lines stick in the same way as the weapon change, one console line only after fast wheel rotation with several clicks.

I guess 1. is the relevant outcome for further investigating, but 2. indicates a general event handling problem, might be related to sticky wheel reaction in 1. or not.

**Comment by zturtleman** [comment]
e.wheel.x and e.wheel.y are both 0 for slow single scroll so ioquake3 doesn't create a K_MWHEELUP or K_MWHEELDOWN event. This is why it doesn't change weapons. It seems to be an SDL2 issue.

**Comment by zturtleman** [comment]
I built the latest SDL and it [fixes the issue](https://hg.libsdl.org/SDL/rev/32b708d2f684). e.wheel.y is never 0 for vertical scrolling. So this should be fixed in ioquake3 after updating to the yet-to-be-released SDL 2.0.9.

**Comment by Velociraptor555** [comment]
The linked libSDL patch looks promising, I will test it immediately upon availability in MacPorts and report the result here.

**Comment by Velociraptor555** [comment]
Today I've compiled libSDL2 using the head revision of the Mercurial repository and linked it against ioquake3: Issue fixed.

**Comment by mcaldas** [comment]
@Velociraptor555 could u help with some info on how to do it?

**Comment by Velociraptor555** [comment]
Pull ioquake3 from git-repo.
Pull libSDL2 from hg-repo.
Configure/build libSDL2 locally.
Replace libSDL2-2.0.0.dylib and libSDL2main.a in ioq3/code/libs/macosx with the new ones.
Build/install ioquake3 as usual.

**Comment by mcaldas** [comment]
Ok, so i didnt manage to build SDL... 
And shouldnt this be closed once is fixed in ioq3?

**Comment by zturtleman** [comment]
Yeah, could leave the issue open until ioq3 updates to SDL 2.0.9.

**Comment by Velociraptor555** [comment]
I've committed libSDL2 #12274 binaries to make things simpler for the time till official release, that's why I closed issue #379. But unfortunately I'm not allowed to git-push on the official repo.

**Comment by mcaldas** [comment]
Maybe u can create a PR with the files?

**Comment by andreagiavatto** [comment]
SDL [2.0.9](https://www.libsdl.org/download-2.0.php) is now out :)

**Comment by Velociraptor555** [comment]
Not in ioq3/code/libs/macosx:

git log libSDL2-2.0.0.dylib 
commit 5bf60a95041bce7b1f42fdf1a4550b57bf83da47
Author: MAN-AT-ARMS <M4N4T4RMS@gmail.com>
Date:   Sat Apr 14 18:49:28 2018 -0400

    Update SDL2 to 2.0.8

**Comment by ensiform** [comment]
Nobody said it was in ioq3 yet.  You can still replace it with your own.

**Comment by BryanWall** [comment]
I added a pull request to update the SDL version for the Mac build to 2.0.10. I tested ioq3 with this build of SDL2 and it fixed the mouse wheel issue. I didn't find any regressions caused by the updated libs.

**Comment by TimeDoctor** [comment]
FYI my current plan (sorry for not communicating this!) is to get an official SDL dylib built by @rcgordon's new build bot box when he has time, and then figure out how to automatically pull that into our repo, so we don't have to do a PR every time there's a new SDL release.

I appreciate everyone who has built a binary themselves, but I have a heightened security awareness now (see https://ioquake3.org/2019/06/13/ioquake3-security-notice-06-13-19-test-builds-possibly-compromised/ ) that makes me less trustful of what are almost certainly fine third-party builds of SDL. I will leave this issue open until a new SDL is in our repo.

**Comment by edonkeycoin** [comment]
Any news on this? I built ioquake from the master branch today and am experiencing the same mouse wheel problem.

Also, an update to the latest SDL built against macOS 10.9 or greater will solve an unrealistic;ated issue that I reported today, so that would be great:

https://github.com/ioquake/ioq3/issues/424

**Comment by rcgordon** [comment]
At this point, this should probably wait for SDL 2.0.12, which is scheduled for the end of October.

**Comment by e-moe** [comment]
still having the same issue :(

**Comment by rcgordon** [comment]
We're like a thousand years behind on SDL 2.0.12.  :(

**Comment by Velociraptor555** [comment]
> still having the same issue :(

On October 2, 2018 I've posted on how to fix it locally, in this thread. At the moment I'm on libSDL 2.0.10 an happily playing every week with my friends.

**Comment by e-moe** [comment]
> > still having the same issue :(
> 
> On October 2, 2018 I've posted on how to fix it locally, in this thread. At the moment I'm on libSDL 2.0.10 an happily playing every week with my friends.

thanks, it's working. I hope to have a new lib in this repository
